src/shared/mount-util.c

```diff
@@ -490,47 +490,77 @@ int mount_move_root(const char *path) {
+                if (fd_newroot < 0)
+                        return log_debug_errno(errno, "Failed to open new rootfs '%s': %m", path);
 
-        fd_newroot = open(path, O_PATH|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
-        if (fd_newroot < 0)
-                return log_debug_errno(errno, "Failed to open new rootfs '%s': %m", path);
+                /* Change into the new rootfs. */
+                if (fchdir(fd_newroot) < 0)
+                        return log_debug_errno(errno, "Failed to change into new rootfs '%s': %m", path);
 
-        /* Change into the new rootfs. */
-        if (fchdir(fd_newroot) < 0)
-                return log_debug_errno(errno, "Failed to change into new rootfs '%s': %m", path);
+                /* Let the kernel tuck the new root under the old one. */
+                if (pivot_root(".", ".") < 0)
+                        return log_debug_errno(errno, "Failed to pivot root to new rootfs '%s': %m", path);
 
-        /* Let the kernel tuck the new root under the old one. */
-        if (pivot_root(".", ".") < 0)
-                return log_debug_errno(errno, "Failed to pivot root to new rootfs '%s': %m", path);
 
+                /* At this point the new root is tucked under the old root. If we want
+                 * to unmount it we cannot be fchdir()ed into it. So escape back to the
+                 * old root. */
+                if (fchdir(fd_oldroot) < 0)
+                        return log_debug_errno(errno, "Failed to change back to old rootfs: %m");
 
-        /* At this point the new root is tucked under the old root. If we want
-         * to unmount it we cannot be fchdir()ed into it. So escape back to the
-         * old root. */
-        if (fchdir(fd_oldroot) < 0)
-                return log_debug_errno(errno, "Failed to change back to old rootfs: %m");
+                /* Note, usually we should set mount propagation up here but we'll
+                 * assume that the caller has already done that. */
 
-        /* Note, usually we should set mount propagation up here but we'll
-         * assume that the caller has already done that. */
+                /* Get rid of the old root and reveal our brand new root. */
+                if (umount2(".", MNT_DETACH) < 0)
+                        return log_debug_errno(errno, "Failed to unmount old rootfs: %m");
 
-        /* Get rid of the old root and reveal our brand new root. */
-        if (umount2(".", MNT_DETACH) < 0)
-                return log_debug_errno(errno, "Failed to unmount old rootfs: %m");
+                if (fchdir(fd_newroot) < 0)
+                        return log_debug_errno(errno, "Failed to switch to new rootfs '%s': %m", path);
+        }
 
-        if (fchdir(fd_newroot) < 0)
-                return log_debug_errno(errno, "Failed to switch to new rootfs '%s': %m", path);
+        /* Finally, let's establish the requested propagation type. */
+        if (flags && mount(NULL, "/", NULL, flags | MS_REC, 0) < 0)
```

Comment:
Please do not use the implicit cast to boolean: `flags != 0`

Reply:
> I've not followed recent private discussions in Telegram, but why not simply add `mount()` with an appropriate flag after `mount_pivot_root()` in nspawn.c?

Conceptually it makes more sense to establish the new rootfs and it's propagation in a single call rather than split it over multiple lines. There's a tight coupling between propagation settings and and chroot/pivot_root et al so I prefer to do it in one place. But I'm not fuzzed if people would like to sprinkle propagation calls into the code directly.